### PR TITLE
[contents,fs.req.namespace] Qualify only namespace-scope names LWG2818

### LINKED ISSUE #1793: I do not think [fs.req.namespaces] means what it thinks it means LWG2818
>  Unless otherwise specified, references to entities described in this subclause are assumed to be qualified with `::std::filesystem::`.

"this subclause" only refers to 30.10.4.1 [fs.req.namespaces], but it really means the whole of 30.10 [filesystems], or at least all the subclauses after 30.4.2.1, not just 30.10.4.1 itself.

Do we really mean _all_ entities? Variables are entities. Class members are entities, and we don't mean to qualify all class members with the namespace name, because e.g. `native()` means `::std::filesystem::path::native()` not `::std::filesystem::native()` (which doesn't exist). I think we mean types and non-member functions. And maybe some typedefs, which are not entities.

20.5.1.1 [contents] specifies a similar rule differently, talking about "a name `x`" not "entities".

We could qualify all uses of types and non-member functions with `filesystem::` so that the general rule in [contents] applies, which might hurt readability, or we could fix [fs.req.namespaces].

### PR DESCRIPTION
**Pull request by jensmaurer**
It does not make sense to attempt to qualify class members
with a nested-name-specifier consisting of namespace-names.

Fixes #1793.

### CODE REVIEW

**Comment by jensmaurer** [comment]
Editorial meeting: What if there are nested namespaces? What if this is ambiguous (different namespaces)?

**Comment by jensmaurer** [comment]
I have rebased the patch, but it does not yet implement the "editorial meeting" outcome. This patch is currently not ready to be merged.

**Comment by zygoloid** [comment]
I would like to make sure that LWG has a chance to look at this and comment on it before it's applied.

**Comment by zygoloid** [comment]
We appear to have quite a few places that fail to satisfy this "unless otherwise specified" rule (these are pre-existing errors, but still):

[optional.swap]'s Table 43 says "calls `swap(*(*this), *rhs)` without saying that ADL is performed
[variant.swap]/2.2 likewise says 'calls `swap(...)`' without saying that ADL is performed, and similarly in /3 and /4
[queue.defn]'s inline definition of `std::queue<T>::swap` is:
```
void swap(queue& q) noexcept(is_nothrow_swappable_v<Container>)
  { using std::swap; swap(c, q.c); }
```
... which is wrong, and likewise for [priqueue.overview] and [stack.defn].
[alg.swap]/2 likewise says `swap(*(first1 + n), *(first2 + n))`, and [alg.swap]/6 says `swap(*a, *b)` with no invocation of ADL.

[container.node.modifiers] likewise uses `swap` without a "use ADL" proviso, but I think that one is actually correct: we don't want user code interfering with `node_handle` swaps, but user types can absolutely be associated classes of that lookup

The definition of `is_swappable_with` deals with this by saying "in an overload-resolution context for swappable values (15.5.3.2)", and looks like the *only* place we get this right other than the definition of the "swappable" requirements and `Swappable` concepts.

All this makes me think that perhaps we should move the "swap is found by ADL" rule into the front matter: "Unless otherwise specified, argument-dependent name lookup is only performed within the standard library for *operator-function-id*s and for the *identifier* `swap`." -- and we should change [container.node.modifiers] to explicitly specify that ADL is not performed. Though in the short term, we could instead update the places listed above to say "in an overload-resolution context for swappable values".

**Comment by jensmaurer** [comment]
I've added the front-matter talk about "swap", which results in fewer changes than talking about ADL everywhere "swap" appears.

@jwakely, could you have LWG look at this?

**Comment by jensmaurer** [comment]
Editorial 2021-04-30: Remove iostreams.tex changes; they are covered by LWG2818.

File: source/iostreams.tex
Comment (jwakely) [inline source/iostreams.tex:11544]: This part is present in the lwg2818 resolution, so not needed in the editorial changes, right?

File: source/lib-intro.tex
Comment (jwakely) [inline source/lib-intro.tex:910]: This part is already being removed by LWG 2818 so I think this editorial change would be clearer with just the addition below (possibly rebased on 2818).

File: source/lib-intro.tex
Comment (jwakely) [inline source/lib-intro.tex:914]: I think it's a bit unclear what "that" refers to here. Maybe "Using a _qualified-id_ avoids argument-dependent lookup." instead?

**Review by jwakely — CHANGES_REQUESTED** [review]

File: source/iostreams.tex
Comment (jensmaurer) [reply source/iostreams.tex:11544]: I think this entire pull request is mostly moot with LWG 2818 applied. I'll wait until that one has landed before reviewing for residual cleanup.

Comment (tkoeppe) [reply source/iostreams.tex:11544]: LWG 2818 has landed!

**Comment by jensmaurer** [comment]
The only residual open issue is the special treatment of `swap`. See the updated diff.

**Comment by jwakely** [comment]
See also [LWG 3629: `make_error_code` and `make_error_condition` are customization points](https://cplusplus.github.io/LWG/issue3629).

File: source/lib-intro.tex
Comment (Quuxplusone) [inline source/lib-intro.tex:952]: On line 943, there's an example for `is_array_v`, which gives help where none is needed (type and variable names never change their lookup results based on qualification anyway); but there is no example given for line 937, which is where help _is_ needed (because it is the function lookup rules that are complicated already, and which are being changed by this wording). I suggest adding a new note, e.g.
```
\begin{note}
The reference to \tcode{swap_ranges} in the specification of \tcode{std::swap}\iref{utility.swap}
refers to \tcode{::std::swap_ranges}.
\end{note}
```

Comment (jwakely) [reply source/lib-intro.tex:952]: Given that `swap` is the special case called out in this wording, using the definition of `std::swap` for an example where the special case doesn't apply can only lead to confusion.

Comment (jwakely) [reply source/lib-intro.tex:952]: And I don't think this is needed anyway. The rule is consistent: unless specified otherwise, we do qualified lookup. Function or not, you do qualified lookup. So the extra complication that applies to lookup for functions (ADL) isn't relevant, we're explicitly not doing that.

Comment (Quuxplusone) [reply source/lib-intro.tex:952]: > can only lead to confusion...

...or clarification. ;)

The rule from the implementor's point of view is "Variable and type names are treated as written (e.g. when the paper standard says `is_same_v`, the implementor should write `is_same_v` in the code). Function names get special treatment (e.g. when the paper standard says `swap_ranges`, the implementor should write `std::swap_ranges` in the code)." The extra complication exists _only_ for function names, and yet that's the one case where no clarifying example is provided.

**Comment by jensmaurer** [comment]
> when the paper standard says swap_ranges, the implementor should write std::swap_ranges in the code

This is not accurate:
```
namespace N {
  namespace std {
    void f();
  }
}

using namespace N;

namespace std {
  void f();
  void g() { std::f(); }
}
```
x.cc: In function ‘void std::g()’:
x.cc:12:14: error: reference to ‘std’ is ambiguous
   12 |   void g() { std::f(); }
      |              ^~~
x.cc:3:13: note: candidates are: ‘namespace N::std { }’
    3 |   namespace std {
      |             ^~~
<built-in>: note:                 ‘namespace std { }’
``` 

So, the implementer needs to use `::std::function_name()` to disambiguate properly.
(If the function is not overloaded or a template, you might also get away with something like `(+f)()` (convert to a function pointer first, which doesn't use ADL).)

File: source/lib-intro.tex
Comment (Quuxplusone) [reply source/lib-intro.tex:952]: @jensmaurer wrote:
> [Actually] the implementer needs to use [`::std::swap_ranges`] to disambiguate properly.

True. To rephrase:

> The rule from the implementor's point of view is "Variable and type names are treated as written. Function names get special treatment." The extra complication exists _only_ for function names, and yet that's the one case where no clarifying example is provided.

Proving that the special treatment of functions is _even more special_ than previously indicated, merely strengthens the force of the top-level statement.

**Pull request merged**
